// itowns/lib/Main.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Extent", {
  enumerable: true,
  get: function get() {
    return _Extent["default"];
  }
});
Object.defineProperty(exports, "Coordinates", {
  enumerable: true,
  get: function get() {
    return _Coordinates["default"];
  }
});
Object.defineProperty(exports, "CRS", {
  enumerable: true,
  get: function get() {
    return _Crs["default"];
  }
});
Object.defineProperty(exports, "Ellipsoid", {
  enumerable: true,
  get: function get() {
    return _Ellipsoid["default"];
  }
});
Object.defineProperty(exports, "ellipsoidSizes", {
  enumerable: true,
  get: function get() {
    return _Ellipsoid.ellipsoidSizes;
  }
});
Object.defineProperty(exports, "GlobeView", {
  enumerable: true,
  get: function get() {
    return _GlobeView["default"];
  }
});
Object.defineProperty(exports, "GLOBE_VIEW_EVENTS", {
  enumerable: true,
  get: function get() {
    return _GlobeView.GLOBE_VIEW_EVENTS;
  }
});
Object.defineProperty(exports, "PlanarView", {
  enumerable: true,
  get: function get() {
    return _PlanarView["default"];
  }
});
Object.defineProperty(exports, "Fetcher", {
  enumerable: true,
  get: function get() {
    return _Fetcher["default"];
  }
});
Object.defineProperty(exports, "MAIN_LOOP_EVENTS", {
  enumerable: true,
  get: function get() {
    return _MainLoop.MAIN_LOOP_EVENTS;
  }
});
Object.defineProperty(exports, "View", {
  enumerable: true,
  get: function get() {
    return _View["default"];
  }
});
Object.defineProperty(exports, "VIEW_EVENTS", {
  enumerable: true,
  get: function get() {
    return _View.VIEW_EVENTS;
  }
});
Object.defineProperty(exports, "Feature", {
  enumerable: true,
  get: function get() {
    return _Feature["default"];
  }
});
Object.defineProperty(exports, "FeatureCollection", {
  enumerable: true,
  get: function get() {
    return _Feature.FeatureCollection;
  }
});
Object.defineProperty(exports, "FeatureGeometry", {
  enumerable: true,
  get: function get() {
    return _Feature.FeatureGeometry;
  }
});
Object.defineProperty(exports, "FEATURE_TYPES", {
  enumerable: true,
  get: function get() {
    return _Feature.FEATURE_TYPES;
  }
});
Object.defineProperty(exports, "process3dTilesNode", {
  enumerable: true,
  get: function get() {
    return _dTilesProcessing.process3dTilesNode;
  }
});
Object.defineProperty(exports, "init3dTilesLayer", {
  enumerable: true,
  get: function get() {
    return _dTilesProcessing.init3dTilesLayer;
  }
});
Object.defineProperty(exports, "$3dTilesCulling", {
  enumerable: true,
  get: function get() {
    return _dTilesProcessing.$3dTilesCulling;
  }
});
Object.defineProperty(exports, "$3dTilesSubdivisionControl", {
  enumerable: true,
  get: function get() {
    return _dTilesProcessing.$3dTilesSubdivisionControl;
  }
});
Object.defineProperty(exports, "pre3dTilesUpdate", {
  enumerable: true,
  get: function get() {
    return _dTilesProcessing.pre3dTilesUpdate;
  }
});
Object.defineProperty(exports, "$3dTilesExtensions", {
  enumerable: true,
  get: function get() {
    return _dTilesProvider.$3dTilesExtensions;
  }
});
Object.defineProperty(exports, "$3dTilesAbstractExtension", {
  enumerable: true,
  get: function get() {
    return _dTilesProvider.$3dTilesAbstractExtension;
  }
});
Object.defineProperty(exports, "FeatureProcessing", {
  enumerable: true,
  get: function get() {
    return _FeatureProcessing["default"];
  }
});
Object.defineProperty(exports, "updateLayeredMaterialNodeImagery", {
  enumerable: true,
  get: function get() {
    return _LayeredMaterialNodeProcessing.updateLayeredMaterialNodeImagery;
  }
});
Object.defineProperty(exports, "updateLayeredMaterialNodeElevation", {
  enumerable: true,
  get: function get() {
    return _LayeredMaterialNodeProcessing.updateLayeredMaterialNodeElevation;
  }
});
Object.defineProperty(exports, "OrientedImageCamera", {
  enumerable: true,
  get: function get() {
    return _OrientedImageCamera["default"];
  }
});
Object.defineProperty(exports, "PointsMaterial", {
  enumerable: true,
  get: function get() {
    return _PointsMaterial["default"];
  }
});
Object.defineProperty(exports, "PointCloudProcessing", {
  enumerable: true,
  get: function get() {
    return _PointCloudProcessing["default"];
  }
});
Object.defineProperty(exports, "FlyControls", {
  enumerable: true,
  get: function get() {
    return _FlyControls["default"];
  }
});
Object.defineProperty(exports, "FirstPersonControls", {
  enumerable: true,
  get: function get() {
    return _FirstPersonControls["default"];
  }
});
Object.defineProperty(exports, "StreetControls", {
  enumerable: true,
  get: function get() {
    return _StreetControls["default"];
  }
});
Object.defineProperty(exports, "PlanarControls", {
  enumerable: true,
  get: function get() {
    return _PlanarControls["default"];
  }
});
Object.defineProperty(exports, "CONTROL_EVENTS", {
  enumerable: true,
  get: function get() {
    return _GlobeControls.CONTROL_EVENTS;
  }
});
Object.defineProperty(exports, "Feature2Mesh", {
  enumerable: true,
  get: function get() {
    return _Feature2Mesh["default"];
  }
});
Object.defineProperty(exports, "FeaturesUtils", {
  enumerable: true,
  get: function get() {
    return _FeaturesUtils["default"];
  }
});
Object.defineProperty(exports, "DEMUtils", {
  enumerable: true,
  get: function get() {
    return _DEMUtils["default"];
  }
});
Object.defineProperty(exports, "CameraUtils", {
  enumerable: true,
  get: function get() {
    return _CameraUtils["default"];
  }
});
Object.defineProperty(exports, "OrientationUtils", {
  enumerable: true,
  get: function get() {
    return _OrientationUtils["default"];
  }
});
Object.defineProperty(exports, "ShaderChunk", {
  enumerable: true,
  get: function get() {
    return _ShaderChunk["default"];
  }
});
Object.defineProperty(exports, "getMaxColorSamplerUnitsCount", {
  enumerable: true,
  get: function get() {
    return _LayeredMaterial.getMaxColorSamplerUnitsCount;
  }
});
Object.defineProperty(exports, "Capabilities", {
  enumerable: true,
  get: function get() {
    return _Capabilities["default"];
  }
});
Object.defineProperty(exports, "Layer", {
  enumerable: true,
  get: function get() {
    return _Layer["default"];
  }
});
Object.defineProperty(exports, "ImageryLayers", {
  enumerable: true,
  get: function get() {
    return _Layer.ImageryLayers;
  }
});
Object.defineProperty(exports, "ColorLayer", {
  enumerable: true,
  get: function get() {
    return _ColorLayer["default"];
  }
});
Object.defineProperty(exports, "ElevationLayer", {
  enumerable: true,
  get: function get() {
    return _ElevationLayer["default"];
  }
});
Object.defineProperty(exports, "GeometryLayer", {
  enumerable: true,
  get: function get() {
    return _GeometryLayer["default"];
  }
});
Object.defineProperty(exports, "TiledGeometryLayer", {
  enumerable: true,
  get: function get() {
    return _TiledGeometryLayer["default"];
  }
});
Object.defineProperty(exports, "OrientedImageLayer", {
  enumerable: true,
  get: function get() {
    return _OrientedImageLayer["default"];
  }
});
Object.defineProperty(exports, "STRATEGY_MIN_NETWORK_TRAFFIC", {
  enumerable: true,
  get: function get() {
    return _LayerUpdateStrategy.STRATEGY_MIN_NETWORK_TRAFFIC;
  }
});
Object.defineProperty(exports, "STRATEGY_GROUP", {
  enumerable: true,
  get: function get() {
    return _LayerUpdateStrategy.STRATEGY_GROUP;
  }
});
Object.defineProperty(exports, "STRATEGY_PROGRESSIVE", {
  enumerable: true,
  get: function get() {
    return _LayerUpdateStrategy.STRATEGY_PROGRESSIVE;
  }
});
Object.defineProperty(exports, "STRATEGY_DICHOTOMY", {
  enumerable: true,
  get: function get() {
    return _LayerUpdateStrategy.STRATEGY_DICHOTOMY;
  }
});
Object.defineProperty(exports, "ColorLayersOrdering", {
  enumerable: true,
  get: function get() {
    return _ColorLayersOrdering["default"];
  }
});
Object.defineProperty(exports, "GlobeLayer", {
  enumerable: true,
  get: function get() {
    return _GlobeLayer["default"];
  }
});
Object.defineProperty(exports, "PlanarLayer", {
  enumerable: true,
  get: function get() {
    return _PlanarLayer["default"];
  }
});
Object.defineProperty(exports, "Source", {
  enumerable: true,
  get: function get() {
    return _Source["default"];
  }
});
Object.defineProperty(exports, "FileSource", {
  enumerable: true,
  get: function get() {
    return _FileSource["default"];
  }
});
Object.defineProperty(exports, "TMSSource", {
  enumerable: true,
  get: function get() {
    return _TMSSource["default"];
  }
});
Object.defineProperty(exports, "WFSSource", {
  enumerable: true,
  get: function get() {
    return _WFSSource["default"];
  }
});
Object.defineProperty(exports, "WMSSource", {
  enumerable: true,
  get: function get() {
    return _WMSSource["default"];
  }
});
Object.defineProperty(exports, "WMTSSource", {
  enumerable: true,
  get: function get() {
    return _WMTSSource["default"];
  }
});
Object.defineProperty(exports, "OrientedImageSource", {
  enumerable: true,
  get: function get() {
    return _OrientedImageSource["default"];
  }
});
Object.defineProperty(exports, "GpxParser", {
  enumerable: true,
  get: function get() {
    return _GpxParser["default"];
  }
});
Object.defineProperty(exports, "GeoJsonParser", {
  enumerable: true,
  get: function get() {
    return _GeoJsonParser["default"];
  }
});
Object.defineProperty(exports, "KMLParser", {
  enumerable: true,
  get: function get() {
    return _KMLParser["default"];
  }
});
Object.defineProperty(exports, "CameraCalibrationParser", {
  enumerable: true,
  get: function get() {
    return _CameraCalibrationParser["default"];
  }
});
Object.defineProperty(exports, "BatchTableHierarchyExtensionParser", {
  enumerable: true,
  get: function get() {
    return _BatchTableHierarchyExtensionParser["default"];
  }
});
Object.defineProperty(exports, "ShapefileParser", {
  enumerable: true,
  get: function get() {
    return _ShapefileParser["default"];
  }
});
Object.defineProperty(exports, "enableDracoLoader", {
  enumerable: true,
  get: function get() {
    return _B3dmParser.enableDracoLoader;
  }
});
Object.defineProperty(exports, "glTFLoader", {
  enumerable: true,
  get: function get() {
    return _B3dmParser.glTFLoader;
  }
});
Object.defineProperty(exports, "legacyGLTFLoader", {
  enumerable: true,
  get: function get() {
    return _B3dmParser.legacyGLTFLoader;
  }
});

var _Extent = _interopRequireDefault(require("./Core/Geographic/Extent"));

var _Coordinates = _interopRequireDefault(require("./Core/Geographic/Coordinates"));

var _Crs = _interopRequireDefault(require("./Core/Geographic/Crs"));

var _Ellipsoid = _interopRequireWildcard(require("./Core/Math/Ellipsoid"));

var _GlobeView = _interopRequireWildcard(require("./Core/Prefab/GlobeView"));

var _PlanarView = _interopRequireDefault(require("./Core/Prefab/PlanarView"));

var _Fetcher = _interopRequireDefault(require("./Provider/Fetcher"));

var _MainLoop = require("./Core/MainLoop");

var _View = _interopRequireWildcard(require("./Core/View"));

var _Feature = _interopRequireWildcard(require("./Core/Feature"));

var _dTilesProcessing = require("./Process/3dTilesProcessing");

var _dTilesProvider = require("./Provider/3dTilesProvider");

var _FeatureProcessing = _interopRequireDefault(require("./Process/FeatureProcessing"));

var _LayeredMaterialNodeProcessing = require("./Process/LayeredMaterialNodeProcessing");

var _OrientedImageCamera = _interopRequireDefault(require("./Renderer/OrientedImageCamera"));

var _PointsMaterial = _interopRequireDefault(require("./Renderer/PointsMaterial"));

var _PointCloudProcessing = _interopRequireDefault(require("./Process/PointCloudProcessing"));

var _FlyControls = _interopRequireDefault(require("./Controls/FlyControls"));

var _FirstPersonControls = _interopRequireDefault(require("./Controls/FirstPersonControls"));

var _StreetControls = _interopRequireDefault(require("./Controls/StreetControls"));

var _PlanarControls = _interopRequireDefault(require("./Controls/PlanarControls"));

var _GlobeControls = require("./Controls/GlobeControls");

var _Feature2Mesh = _interopRequireDefault(require("./Converter/Feature2Mesh"));

var _FeaturesUtils = _interopRequireDefault(require("./Utils/FeaturesUtils"));

var _DEMUtils = _interopRequireDefault(require("./Utils/DEMUtils"));

var _CameraUtils = _interopRequireDefault(require("./Utils/CameraUtils"));

var _OrientationUtils = _interopRequireDefault(require("./Utils/OrientationUtils"));

var _ShaderChunk = _interopRequireDefault(require("./Renderer/Shader/ShaderChunk"));

var _LayeredMaterial = require("./Renderer/LayeredMaterial");

var _Capabilities = _interopRequireDefault(require("./Core/System/Capabilities"));

var _Layer = _interopRequireWildcard(require("./Layer/Layer"));

var _ColorLayer = _interopRequireDefault(require("./Layer/ColorLayer"));

var _ElevationLayer = _interopRequireDefault(require("./Layer/ElevationLayer"));

var _GeometryLayer = _interopRequireDefault(require("./Layer/GeometryLayer"));

var _TiledGeometryLayer = _interopRequireDefault(require("./Layer/TiledGeometryLayer"));

var _OrientedImageLayer = _interopRequireDefault(require("./Layer/OrientedImageLayer"));

var _LayerUpdateStrategy = require("./Layer/LayerUpdateStrategy");

var _ColorLayersOrdering = _interopRequireDefault(require("./Renderer/ColorLayersOrdering"));

var _GlobeLayer = _interopRequireDefault(require("./Core/Prefab/Globe/GlobeLayer"));

var _PlanarLayer = _interopRequireDefault(require("./Core/Prefab/Planar/PlanarLayer"));

var _Source = _interopRequireDefault(require("./Source/Source"));

var _FileSource = _interopRequireDefault(require("./Source/FileSource"));

var _TMSSource = _interopRequireDefault(require("./Source/TMSSource"));

var _WFSSource = _interopRequireDefault(require("./Source/WFSSource"));

var _WMSSource = _interopRequireDefault(require("./Source/WMSSource"));

var _WMTSSource = _interopRequireDefault(require("./Source/WMTSSource"));

var _OrientedImageSource = _interopRequireDefault(require("./Source/OrientedImageSource"));

var _GpxParser = _interopRequireDefault(require("./Parser/GpxParser"));

var _GeoJsonParser = _interopRequireDefault(require("./Parser/GeoJsonParser"));

var _KMLParser = _interopRequireDefault(require("./Parser/KMLParser"));

var _CameraCalibrationParser = _interopRequireDefault(require("./Parser/CameraCalibrationParser"));

var _BatchTableHierarchyExtensionParser = _interopRequireDefault(require("./Parser/BatchTableHierarchyExtensionParser"));

var _ShapefileParser = _interopRequireDefault(require("./Parser/ShapefileParser"));

var _B3dmParser = require("./Parser/B3dmParser");
