// itowns/lib/Scene/Description/StyleManager.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _create = require('babel-runtime/core-js/object/create');

var _create2 = _interopRequireDefault(_create);

var _Style = require('./Style');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function StyleManager() {
    this.styles = [];
} /**
   * Generated On: 2015-10-5
   * Class: StyleManager
   * Description:
   */

StyleManager.prototype = (0, _create2.default)(StyleManager.prototype);

StyleManager.prototype.constructor = StyleManager;

StyleManager.prototype.getStyles = function getStyles() {
    return this.styles;
};

StyleManager.prototype.addStyle = function addStyle(style) {
    if (style instanceof _Style.Style) {
        this.styles.push(style);
    }
};

StyleManager.prototype.getStyle = function getStyle(idStyle) {
    return this.styles.filter(function (element) {
        return element.name === idStyle;
    }).pop();
};

StyleManager.prototype.removeStyle = function removeStyle(idStyle) {
    this.styles = this.styles.filter(function (element) {
        return element.name === idStyle;
    });
};

StyleManager.prototype.updateStyle = function updateStyle() {};

exports.default = StyleManager;
