// itowns/lib/Globe/TileMesh.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _getIterator2 = require('babel-runtime/core-js/get-iterator');

var _getIterator3 = _interopRequireDefault(_getIterator2);

var _create = require('babel-runtime/core-js/object/create');

var _create2 = _interopRequireDefault(_create);

var _three = require('three');

var THREE = _interopRequireWildcard(_three);

var _NodeMesh = require('../Renderer/NodeMesh');

var _NodeMesh2 = _interopRequireDefault(_NodeMesh);

var _LayeredMaterial = require('../Renderer/LayeredMaterial');

var _LayeredMaterial2 = _interopRequireDefault(_LayeredMaterial);

var _GlobeDepthMaterial = require('../Renderer/GlobeDepthMaterial');

var _GlobeDepthMaterial2 = _interopRequireDefault(_GlobeDepthMaterial);

var _MatteIdsMaterial = require('../Renderer/MatteIdsMaterial');

var _MatteIdsMaterial2 = _interopRequireDefault(_MatteIdsMaterial);

var _RendererConstant = require('../Renderer/RendererConstant');

var _RendererConstant2 = _interopRequireDefault(_RendererConstant);

var _OGCWebServiceHelper = require('../Core/Commander/Providers/OGCWebServiceHelper');

var _OGCWebServiceHelper2 = _interopRequireDefault(_OGCWebServiceHelper);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function TileMesh(geometry, params) {
    // Constructor
    _NodeMesh2.default.call(this);

    this.matrixAutoUpdate = false;
    this.rotationAutoUpdate = false;

    if (!params.bbox) {
        throw new Error('params.bbox is mandatory to build a TileMesh');
    }

    this.level = params.level;
    this.bbox = params.bbox;

    this.geometry = geometry;
    this.normal = params.center.clone().normalize();

    // TODO Why move sphere center
    this.centerSphere = new THREE.Vector3().addVectors(this.geometry.boundingSphere.center, params.center);

    this.oSphere = new THREE.Sphere(this.centerSphere.clone(), this.geometry.boundingSphere.radius);

    this.materials = [];

    if (params.parentMaterial instanceof _LayeredMaterial2.default) {
        if (params.parentMaterial.getElevationLayerLevel() > -1) {
            _OGCWebServiceHelper2.default.computeTileMatrixSetCoordinates(this);
        }
        for (var i = params.parentMaterial.colorLayersId.length - 1; i >= 0; i--) {
            var projection = params.parentMaterial.getLayerUV(i);
            _OGCWebServiceHelper2.default.computeTileMatrixSetCoordinates(this, projection ? 'PM' : 'WGS84G');
        }
    }
    // instantiations all state materials : final, depth, id
    // Final rendering : return layered color + fog
    this.materials[_RendererConstant2.default.FINAL] = new _LayeredMaterial2.default(params.parentMaterial, this.wmtsCoords, params.parentWmtsCoords);

    // Depth : return the distance between projection point and the node
    this.materials[_RendererConstant2.default.DEPTH] = new _GlobeDepthMaterial2.default(this.materials[_RendererConstant2.default.FINAL]);
    // ID : return id color in RGBA (float Pack in RGBA)
    this.materials[_RendererConstant2.default.ID] = new _MatteIdsMaterial2.default(this.materials[_RendererConstant2.default.FINAL]);
    // Set current material in Final Rendering
    this.material = this.materials[_RendererConstant2.default.FINAL];

    this.frustumCulled = false;

    // Layer
    this.setDisplayed(false);
} /**
   * Generated On: 2015-10-5
   * Class: TileMesh
   * Description: Tuile de maillage, noeud du quadtree MNT. Le Materiel est issus du QuadTree ORTHO.
   */

TileMesh.prototype = (0, _create2.default)(_NodeMesh2.default.prototype);

TileMesh.prototype.constructor = TileMesh;

TileMesh.prototype.dispose = function dispose() {
    // TODO à mettre dans node mesh
    this.material.dispose();
    this.geometry.dispose();
    this.geometry = null;
    this.material = null;
};

TileMesh.prototype.setUuid = function setUuid(uuid) {
    this.id = uuid;
    this.materials[_RendererConstant2.default.FINAL].setUuid(uuid);
    this.materials[_RendererConstant2.default.ID].setUuid(uuid);
};

TileMesh.prototype.getUuid = function getUuid(uuid) {
    return this.materials[_RendererConstant2.default.ID].getUuid(uuid);
};

TileMesh.prototype.setLightingParameters = function setLightingParameters(lighting) {
    var material = this.materials[_RendererConstant2.default.FINAL];
    material.setLightingOn(lighting.enable);
    material.uniforms.lightPosition.value = lighting.position;
};
/**
 *
 * @returns {undefined}     */
TileMesh.prototype.disposeChildren = function disposeChildren() {
    this.pendingSubdivision = false;

    while (this.children.length > 0) {
        var child = this.children[0];
        this.remove(child);
        child.dispose();
        if (child.content) {
            var content = child.content;
            content.forEach(function (object) {
                if (object.parent) {
                    object.parent.remove(object);
                }
                for (var i = object.children.length - 1; i >= 0; i--) {
                    object.children[i].geometry.dispose();
                    object.children[i].material.dispose();
                }
            });
        }
    }
};

TileMesh.prototype.setDisplayed = function setDisplayed(show) {
    var _iteratorNormalCompletion = true;
    var _didIteratorError = false;
    var _iteratorError = undefined;

    try {
        for (var _iterator = (0, _getIterator3.default)(this.materials), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
            var material = _step.value;

            material.visible = show;
        }
    } catch (err) {
        _didIteratorError = true;
        _iteratorError = err;
    } finally {
        try {
            if (!_iteratorNormalCompletion && _iterator.return) {
                _iterator.return();
            }
        } finally {
            if (_didIteratorError) {
                throw _iteratorError;
            }
        }
    }

    if (show) {
        this.content.forEach(function (element) {
            element.visible = true;
        });
    }
};

TileMesh.prototype.enableRTC = function enableRTC(enable) {
    this.materials[_RendererConstant2.default.FINAL].enableRTC(enable);
};

// switch material in function of state
TileMesh.prototype.changeState = function changeState(state) {
    if (state !== _RendererConstant2.default.FINAL) {
        this.materials[state].visible = this.materials[_RendererConstant2.default.FINAL].visible;
    }

    this.material = this.materials[state];
};

TileMesh.prototype.setFog = function setFog(fog) {
    this.materials[_RendererConstant2.default.FINAL].setFogDistance(fog);
};

TileMesh.prototype.setMatrixRTC = function setMatrixRTC(rtc) {
    var _iteratorNormalCompletion2 = true;
    var _didIteratorError2 = false;
    var _iteratorError2 = undefined;

    try {
        for (var _iterator2 = (0, _getIterator3.default)(this.materials), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
            var material = _step2.value;

            material.setMatrixRTC(rtc);
        }
    } catch (err) {
        _didIteratorError2 = true;
        _iteratorError2 = err;
    } finally {
        try {
            if (!_iteratorNormalCompletion2 && _iterator2.return) {
                _iterator2.return();
            }
        } finally {
            if (_didIteratorError2) {
                throw _iteratorError2;
            }
        }
    }
};

TileMesh.prototype.setDebug = function setDebug(enable) {
    this.materials[_RendererConstant2.default.FINAL].setDebug(enable);
};

TileMesh.prototype.setSelected = function setSelected(select) {
    this.materials[_RendererConstant2.default.FINAL].setSelected(select);
};

TileMesh.prototype.setTextureElevation = function setTextureElevation(elevation) {
    if (this.materials[_RendererConstant2.default.FINAL] === null) {
        return;
    }

    var offsetScale = elevation.pitch || new THREE.Vector3(0, 0, 1);
    this.setBBoxZ(elevation.min, elevation.max);

    this.materials[_RendererConstant2.default.FINAL].setTexture(elevation.texture, _LayeredMaterial.l_ELEVATION, 0, offsetScale);
    this.materials[_RendererConstant2.default.DEPTH].uniforms.texturesCount.value = this.materials[_RendererConstant2.default.FINAL].loadedTexturesCount[0];
    this.materials[_RendererConstant2.default.ID].uniforms.texturesCount.value = this.materials[_RendererConstant2.default.FINAL].loadedTexturesCount[0];
};

TileMesh.prototype.setBBoxZ = function setBBoxZ(min, max) {
    if (Math.floor(min) !== Math.floor(this.bbox.bottom()) || Math.floor(max) !== Math.floor(this.bbox.top())) {
        this.bbox.setBBoxZ(min, max);
        var delta = this.geometry.OBB.addHeight(this.bbox);
        var trans = this.normal.clone().setLength(delta.y);

        this.geometry.boundingSphere.radius = Math.sqrt(delta.x * delta.x + this.oSphere.radius * this.oSphere.radius);
        this.centerSphere = new THREE.Vector3().addVectors(this.oSphere.center, trans);
    }
};

TileMesh.prototype.setTexturesLayer = function setTexturesLayer(textures, layerType, layerId) {
    if (this.material === null) {
        return;
    }
    if (textures) {
        this.material.setTexturesLayer(textures, layerType, layerId);
    }
};

TileMesh.prototype.setTextureFeatures = function setTextureFeatures(texture) {
    this.material.uniforms.rasterFeatures.value = true;
    this.material.uniforms.featureTexture.value = texture;
};

TileMesh.prototype.getLayerTextures = function getLayerTextures(layerType, layerId) {
    var mat = this.materials[_RendererConstant2.default.FINAL];
    return mat.getLayerTextures(layerType, layerId);
};

TileMesh.prototype.isColorLayerLoaded = function isColorLayerLoaded(layerId) {
    var mat = this.materials[_RendererConstant2.default.FINAL];
    return mat.getColorLayerLevelById(layerId) > -1;
};

TileMesh.prototype.isElevationLayerLoaded = function isElevationLayerLoaded() {
    var mat = this.materials[_RendererConstant2.default.FINAL];
    return mat.getElevationLayerLevel() > -1;
};

TileMesh.prototype.isColorLayerDownscaled = function isColorLayerDownscaled(layerId) {
    var mat = this.materials[_RendererConstant2.default.FINAL];
    return mat.isColorLayerDownscaled(layerId, this.wmtsCoords);
};

TileMesh.prototype.isLayerTypeDownscaled = function isLayerTypeDownscaled(layerType) {
    var mat = this.materials[_RendererConstant2.default.FINAL];
    return mat.isLayerTypeDownscaled(layerType, this.wmtsCoords);
};

TileMesh.prototype.normals = function normals() {
    return this.geometry.normals;
};

TileMesh.prototype.fourCorners = function fourCorners() {
    return this.geometry.fourCorners;
};

TileMesh.prototype.normal = function normal() {
    return this.geometry.normal;
};

TileMesh.prototype.center = function center() {
    return this.geometry.center;
};

TileMesh.prototype.OBB = function OBB() {
    return this.geometry.OBB;
};

TileMesh.prototype.getIndexLayerColor = function getIndexLayerColor(idLayer) {
    return this.materials[_RendererConstant2.default.FINAL].indexOfColorLayer(idLayer);
};

TileMesh.prototype.removeColorLayer = function removeColorLayer(idLayer) {
    this.materials[_RendererConstant2.default.FINAL].removeColorLayer(idLayer);
};

TileMesh.prototype.changeSequenceLayers = function changeSequenceLayers(sequence) {
    var layerCount = this.materials[_RendererConstant2.default.FINAL].getColorLayersCount();

    // Quit if there is only one layer
    if (layerCount < 2) {
        return;
    }

    this.materials[_RendererConstant2.default.FINAL].setSequence(sequence);
};

exports.default = TileMesh;
