// itowns/lib/MobileMapping/GeometryProj.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _create = require('babel-runtime/core-js/object/create');

var _create2 = _interopRequireDefault(_create);

var _three = require('three');

var THREE = _interopRequireWildcard(_three);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function GeometryProj() {
    // Constructor

} /**
   *
   * Class: GeometryProj
   */

GeometryProj.prototype = (0, _create2.default)(THREE.BufferGeometry.prototype);
GeometryProj.prototype.constructor = GeometryProj;

GeometryProj.prototype.enableRTC = function enableRTC(enable) {
    this.material.enableRTC(enable);
};

GeometryProj.prototype.setFog = function setFog(fog) {
    this.material.setFogDistance(fog);
};

GeometryProj.prototype.setMatrixRTC = function setMatrixRTC(rtc) {
    this.material.setMatrixRTC(rtc);
};

GeometryProj.prototype.setDebug = function setDebug(enable) {
    this.material.setDebug(enable);
};

GeometryProj.prototype.setSelected = function setSelected(select) {
    this.material.setSelected(select);
};

exports.default = GeometryProj;
